validator/src/main/java/com/google/daq/mqtt/validator/Validator.java

```diff
@@ -632,16 +632,44 @@ private JsonSchema getSchema(File schemaFile) {
 
   private void validateFile(
       String prefix, String targetFile, String schemaName, JsonSchema schema) {
+    final File targetOut = getTargetPath(prefix, targetFile.replace(".json", ".out"));
     try {
       File fullPath = getFullPath(prefix, new File(targetFile));
       Map<String, Object> message = OBJECT_MAPPER.readValue(fullPath, Map.class);
       sanitizeMessage(schemaName, message);
-      validateJsonNode(schema, OBJECT_MAPPER.valueToTree(message));
+      JsonNode jsonNode = OBJECT_MAPPER.valueToTree(message);
+      upgradeMessage(schemaName, jsonNode);
+      OBJECT_MAPPER.writeValue(getTargetPath(prefix, targetFile), jsonNode);
+      validateJsonNode(schema, jsonNode);
+      writeExceptionOutput(targetOut, null);
     } catch (Exception e) {
+      writeExceptionOutput(targetOut, e);
       throw new RuntimeException("Against input " + targetFile, e);
     }
   }
 
+  private void writeExceptionOutput(File targetOut, Exception e) {
+    try (OutputStream outputStream = new FileOutputStream(targetOut)) {
+      if (e != null) {
+        ExceptionMap.format(e, ERROR_FORMAT_INDENT).write(outputStream);
+      }
+    } catch (Exception e2) {
+      throw new RuntimeException("While writing " + targetOut, e2);
+    }
+  }
+
+  private File getTargetPath(String baseFile, String addedFile) {
+    File prefix = new File(baseFile);
+    File outBase = new File(prefix.getParentFile(), "out/tests");
+    File full = new File(outBase, addedFile);
+    full.getParentFile().mkdirs();
+    return full;
+  }
+
+  private void upgradeMessage(String schemaName, JsonNode jsonNode) {
+    new MessageUpgrader(schemaName, jsonNode).upgrade();
```

Comment:
Do we want more robust handling of messages that don't upgrade?

Reply:
After the upgrade it's run through the current schema validator to make sure the upgraded message is proper, so if "didn't upgrade" means that there are still lingering fields or something it will be caught. I'm not sure what additional handling would be useful... might depend on the specifics of the "didn't upgrade" (like, missing parameter, wrong parameters, etc...).

Reply:
I think that "its run through the current schema validator" is at least consistent enough behavior for the user so I'm fine with that.